apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/Keys.cs

```diff
@@ -275,49 +255,56 @@ private V1.KeyRange ToProtobuf()
     /// </summary>
     public sealed class Key
     {
-        /// <summary>
-        /// The values of this key.
-        /// </summary>
-        internal ListValue KeyParts { get; }
+        internal SpannerParameterCollection Parameters { get; }
 
         /// <summary>
         /// Creates a key from the values in the parameter collection.
         /// </summary>
         /// <param name="keyParts">The values that define the key</param>
         internal Key(SpannerParameterCollection keyParts)
         {
-            KeyParts = ToListValue(keyParts);
+            Parameters = keyParts;
         }
 
         /// <summary>
         /// Creates a key consisting of the given key parts.
         /// </summary>
         /// <param name="keyParts">The values that define the key</param>
-        public Key(params object[] keyParts)
+        public Key(params object[] keyParts) : this(new SpannerParameterCollection(keyParts.Select(j => new SpannerParameter { Value = j })))
         {
-            var protobufValues = keyParts.Select(
-                v => SpannerDbType.FromClrType(v?.GetType())?.ToProtobufValue(v) ?? Value.ForNull());
-            KeyParts = new ListValue { Values = { protobufValues } };
         }
 
-        /// <inheritdoc/>
-        public override bool Equals(object other)
-            => other is Key key && KeyParts.Equals(key.KeyParts);
+        /// <summary>
+        /// Gets the values of this Key.
+        /// </summary>
+        /// <param name="options">The options for conversion between protobuf and CLR types.</param>
+        /// <returns>The values of this Key.</returns>
+        internal ListValue GetKeyParts(SpannerConversionOptions options) => ToProtobuf(options);
 
-        /// <inheritdoc/>
-        public override int GetHashCode() => KeyParts.GetHashCode();
+        /// <summary>
+        /// Returns the string representation of this key using the default type conversion mappings.
+        /// The string representation is only for diagnostic purposes and the implementation may change over time.
+        /// </summary>
+        /// <returns>A string that represents the current key.</returns>
+        public override string ToString() => ToProtobuf(SpannerConversionOptions.Default).ToString();
 
-        /// <inheritdoc/>
-        public override string ToString() => KeyParts.ToString();
+        /// <summary>
+        /// Returns the string representation of this key using the type conversion options specified in 
+        /// <see cref="SpannerConnectionStringBuilder"/>. The string representation is only for diagnostic purposes
+        /// and the implementation may change over time.
+        /// </summary>
+        /// <param name="builder">The <see cref="SpannerConnectionStringBuilder"/> instance.</param>
+        /// <returns>A string that represents the current key.</returns>
+        public string ToString(SpannerConnectionStringBuilder builder) => ToProtobuf(builder.ConversionOptions).ToString();
```

Comment:
This one we can keep for diagnostic purposes. Could we should pass ConversionOptions instead?